File: assignment-2-dictionary-python-master/dictionary.py
'''
dictionary.py

This module implements a dictionary of words and definitions.
There there can be multiple definitions for a word. This
implementation uses data classes for the _DictionaryEntry
and _Dictionary types

Created on May 23, 2019

@author: philip gust
'''

'''
Dictionary entry with word and definition
'''
class _DictionaryEntry:
    word: str = None
    definition: str = None

'''
Dictionary with array of _Dictionary entries
'''
class _Dictionary:
    entries: list = []  # empty list of dictionary entries

'''
The dictionary
'''
_dict = _Dictionary()

'''
Return the number of entries in the dictionary

@return number of entries in the dictionary
'''
def getDictionarySize():
    return len(_dict.entries)

''' 
Get dictionary word.

@param entry_no the index of the entry
@return word if entry found else None
'''
def getDictionaryWord(entry_no):
    if entry_no >= 0 and entry_no < len(_dict.entries):
        return _dict.entries[entry_no].word

''' 
Get dictionary definition.

@param entry_no the index of the entry
@return definition if entry found else None
'''
def getDictionaryDefinition(entry_no):
    if entry_no >= 0 and entry_no < len(_dict.entries):
        return _dict.entries[entry_no].definition


''' 
Find dictionary entry for a word. If word ends
with wildcard (*), finds any matching word.

@param word the word to match
@param entry_no the starting entry number
@return entry index or -1 if not found
'''
def getDictionaryEntry(word, entry_no = 0):
    for i in range(entry_no, len(_dict.entries)):
        if word[-1] == "*": # wildcard match
            if _dict.entries[i].word.startswith(word[0:-1]):
                return i
        elif _dict.entries[i].word == word: # regular match
            return i
    return -1

'''
Put dictionary definition entry for word.
@param word the entry word
@param def the entry definition
@return index of new entry
'''
def putDictionaryEntry(word, definition):
    entry = _DictionaryEntry()
    entry.word = word;
    entry.definition = definition
    _dict.entries.append(entry)
    return len(_dict.entries)-1

File: assignment-2-rooneyz-master-2/dictionary.c
/*
 * dictionary.c
 *
 * This file implements a dictionary of words and definitions.
 * There there can be multiple definitions for a word. A word
 * is limited to a length of MAX_WORD characters including the
 * terminating '\0' character, and a definition is limited to
 * MAX_DEF characters including the terminating '\0' character.
 * The dictionary is limited to MAX_ENTRIES entries.
 *
 *  @since 2019-05-09
 *  @author Philip Gust
 *
 *  NOTE: This filed was modified to use a temp file to store
 *  the dictionary definitions instead of using Dictionary struct
 *  Author: Zach Rooney
 *  Date: 2019-05-21
 */

#include <stdlib.h>
#include <stdio.h>
#include <string.h>
#include <assert.h>

#include "dictionary.h"

/** Dictionary entry with word and definition */
struct DictionaryEntry {
	/** word string */
	char word[MAX_WORD];

	/** offset from beginning of file */
	long offset;

	/** length of definition */
	size_t length;


};

/** Dictionary array of entries */
struct Dictionary {
	/** Number of dictionary entries */
	int n_entries;

	/** Dictionary entries */
	struct DictionaryEntry entries[MAX_ENTRIES];

	/** Dictionary file */
	FILE* dFile;
};

/** The dictionary */
static struct Dictionary dictionary;

/**
 * Return the number of entries in the dictionary.
 * @return the number of dictionary entries
 */
int getDictionarySize() {
	return dictionary.n_entries;
}

/**
 * Get dictionary definition.
 *
 * @param name the name to find
 * @param word the definition for the name
 * @return true if entry found
 */
bool getDictionaryWord(int entry, char word[]) {
	if (entry < 0 || entry >= dictionary.n_entries) {
		return false;
	}
	strcpy(word, dictionary.entries[entry].word);
	return true;
}

/**
 * Get dictionary definition.
 *
 * @param name the name to find
 * @param def the definition for the entry
 * @return true if entry found
 */
bool getDictionaryDefinition(int entry, char def[]) {
	if (entry < 0 || entry >= dictionary.n_entries) {
		return false;
	}

	fseek(dictionary.dFile, dictionary.entries[entry].offset, SEEK_SET);
	fread(def, sizeof(char), dictionary.entries[entry].length, dictionary.dFile);
	def [dictionary.entries[entry].length] = '\0';
	return true;
}

/**
 * Find dictionary entry for a word. If word ends
 * with wildcard (*), finds any matching word.
 *
 * @param word the word to match
 * @param start_entry the starting entry
 * @return entry index or -1 if not found
 */
int getDictionaryEntry(const char word[], int start_entry) {
	int wordlen = strlen(word);
	if (wordlen > 0 && start_entry >= 0) {
		for (int entry = start_entry; entry < dictionary.n_entries; entry++) {
			if (word[wordlen-1] == '*') {  // wildcard match
				if (strncmp(word, dictionary.entries[entry].word, wordlen-1) == 0) {
					return entry;
				}
			} else if (strcmp(word, dictionary.entries[entry].word) == 0) {
				return entry;
			}
		}
	}
	return -1;
}

/**
 * Put definition entry for name. Assumes unique entry name.
 *
 * @param word the entry word
 * @param def the entry definition
 * @return index of new entry or -1 if dictionary is full
 */
int putDictionaryEntry(const char word[], const char def[]) {
	size_t tLen = strlen(word);
	size_t tDefLen = strlen(def);
	assert(tLen < MAX_WORD);
	assert(strlen(def) < MAX_DEF);

	if (dictionary.n_entries >= MAX_ENTRIES) {
		return -1;
	}

	if (dictionary.dFile == NULL){
		dictionary.dFile = tmpfile();
	}
	strcpy(dictionary.entries[dictionary.n_entries].word, word);
	fseek(dictionary.dFile, 0, SEEK_END);
	dictionary.entries[dictionary.n_entries].offset = ftell(dictionary.dFile);
	dictionary.entries[dictionary.n_entries].length = tDefLen;
	fwrite(def, sizeof(char), tDefLen, dictionary.dFile);

	return dictionary.n_entries++;
}
